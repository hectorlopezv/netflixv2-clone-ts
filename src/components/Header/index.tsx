/* eslint-disable react/jsx-pascal-case */
import React, {useState} from 'react';
import {Container, 
    Header_ButtonLink as HeaderLink, 
    Frame, 
    Logo, 
    Background,
    Feature,
    Text,
    Profile,
    Dropdown,
    GroupPlayButton,
    Picture,
    FeatureCallOut,
    Group,
    Search,
    PlayButton,
    SearchIcon,
    SearchInput,
    Link as Link_,
} from './styles/Header';
import {
    BrowserRouter as Router,
    Switch,
    Route,
    Link
  } from "react-router-dom";

export interface HeaderProps {
    bg?: boolean;
    children: any;
    src?: string;
    dontShowSmallVp?: any;
}
 
const Header: React.FC<HeaderProps> = ({bg =  true, children, ...props}) => {
    return (bg? <Background {...props}  >{children}</Background> : children);
}
 
export default Header;

export interface Header_FrameProps {
    
}
 
export const Header_Frame: React.FC<Header_FrameProps> = ({children, ...props}) => {
    return ( 
        <Container {...props}>{children}</Container>
     );
}


export interface Group_ {
    
}

export const Header_Group: React.FC<Group_> = ({children, ...props}) => {
    return ( 
        <Group {...props}>{children}</Group>
     );
}


 


export interface Header_LogoProps {
    to?: string;
    alt: string;
    src: string;
}
 
export const Header_Logo: React.FC<Header_LogoProps> = ({to= '/', ...props}) => {
    return ( 
        <Link to={to} >
            <Logo {...props} />
        </Link>
        
     );
}
 
export interface Header_ButtonProps {
    to: string;
}
 
export const Header_ButtonLink: React.FC<Header_ButtonProps> = ({to='/', children, ...props}) => {
    return (  
            <HeaderLink {...props} to={to} style={{ textDecoration: 'none' }}>
                {children}
            </HeaderLink> 
    );
}
 
export interface Header_FeatureProps {
    
}
 
export const Header_Feature: React.FC<Header_FeatureProps> = ({children, ...props}) => {
    return (  
        <Feature {...props}>{children}</Feature>
    );
}
 
export interface Header_TextProps {
    
}
 
export const Header_Text: React.FC<Header_TextProps> = ({children, ...props}) => {
    return (  <Text {...props}>{children}</Text>);
};
 
export interface Header_FeatureCallOutProps {
    
}
 
export const Header_FeatureCallOut: React.FC<Header_FeatureCallOutProps> = ({children, ...props}) => {
    return (  <FeatureCallOut {...props}>{children}</FeatureCallOut>);
}
 
export interface Header_TextLinkProps {
    onClick?: any;
}
 
export const Header_TextLink: React.FC<Header_TextLinkProps> = ({children, ...props}) => {
    return (  <Link_ {...props}>{children}</Link_> );
}
 
export interface Header_ProfileProps {
    
}
 
export const Header_Profile: React.FC<Header_ProfileProps> = ({children, ...props}) => {
    return ( <Profile {...props}>{children}</Profile> );
}
 

export interface Header_PictureProps {
    src: string;
}
 
export const Header_Picture: React.FC<Header_PictureProps> = ({src, children, ...props}) => {
    return ( <Picture src={`/images/users/${src}.png`} {...props}>{children}</Picture> );
}
 

export interface Header_DropdownProps {
    
}
 
export const Header_Dropdown: React.FC<Header_DropdownProps> = ({children, ...props}) => {
    return ( <Dropdown {...props}>{children}</Dropdown> );
}


export interface Header_SearchProps {
    searchTerm: any
    setsearchTerm: any
}
 
export const Header_Search: React.FC<Header_SearchProps> = ({searchTerm, setsearchTerm, ...props}) => {
    //esta activo o no el searching functionality de pusher para hacer algo
    //if ifa active put a image like netflix
    const [searchActive, setsearchActive] = useState(false);

    return (
        <Search {...props}>
            <SearchIcon onClick={() => setsearchActive((prevState) => !prevState)}>
                <img src="/images/icons/search.png" alt="Search" />
            </SearchIcon>
            <SearchInput value={searchTerm} 
                placeholder="Search films and series"
                active={searchActive}
                onChange={(event: any) => setsearchTerm(event.target.value)}/>
        </Search>
    );
}
 
export interface Header_PlayButtonProps {
    
}
 
export const Header_PlayButton: React.FC<Header_PlayButtonProps> = ({children, ...props}) => {
    return (<PlayButton {...props}>{children}</PlayButton>  );
}
 
export interface Header_GroupPlayProps {
    
}
 
export const Header_GroupPlay: React.FC<Header_GroupPlayProps> = ({children, ...props}) => {
    return ( <GroupPlayButton {...props}>{children}</GroupPlayButton> );
}
 
