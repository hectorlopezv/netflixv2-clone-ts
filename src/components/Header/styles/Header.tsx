import styled from 'styled-components/macro';
import {Link as ReactRouterLink} from 'react-router-dom';
import { FaPlay } from "react-icons/fa";

export const Container = styled.div`
    display: flex;
    margin: 0 59px;
    height: 64px;
    padding: 18px 0;
    justify-content: space-between;
    align-items: center;

    a{
        display: flex;
    }

    @media ( max-width: 1000px){
        margin: 0 30px;

    }
`;

export const Logo = styled.img`
    height: 32px;
    width: 108px;
    margin-right: 40px;
    @media (min-width: 1449px) {
        height: 45px;
        width: 167px;
    }
`;



export const Frame = styled.div`

`;

interface backgroundProps {
    src?: string;
    dontShowSmallVp?: boolean;
}
export const Background = styled.div<backgroundProps>`
    display: flex;
    flex-direction: column;
    background: linear-gradient(to bottom, rgba(0, 0, 0, 0.35), rgba(0, 0, 0, 0.1), rgba(0, 0, 0, 0.35)), url(${({ src }) => (src ? `../images/misc/${src}.jpg` : '../images/misc/home-bg.jpg')}) top left / cover
    no-repeat;

    @media (max-width: 1100px){
        ${({dontShowSmallVp}) => dontShowSmallVp && 'background: none;'}
    }


`;


//aply style to react route link
export const Header_ButtonLink = styled<any>(ReactRouterLink)`
    display: block;
    background-color: #e50914;
    width: 84px;
    height: fit-content;
    color: white;
    border: 0;
    font-size: 18px;
    border-radius: 3px;
    padding: 8px 17px;
    cursor: pointer;

    &:focus, &:hover, &:visited, &:link, &:active {
        text-decoration: none;
    }

    &:hover {
    background: #f40612;
    }
`;

//create copy of container and add new styles or overwrite
export const Feature = styled(Container)`
    padding: 150px 0 500px 0;
    flex-direction: column;
    align-items: normal;
    width: 50%;

    @media(max-width: 1100px){
        display: none;
    }
`;


export const Text = styled.p`
   color: white;
   font-size: 22px;
   line-height: normal;
   text-shadow: 2px 2px 4px rgba(0,0,0,.45); 
   margin: 0;
`;

export const FeatureCallOut = styled.h2`
    color: white;
    font-size: 50px;
    line-height: normal;
    font-weight: bold;
    padding: 5px 10px;
    text-shadow: 2px 2px 4px rgba(0,0,0,.45); 
   
    margin-bottom: 20px;
`;

interface LinkProps {
    active?: any;
}
export const Link = styled.p<LinkProps>`
    color: white;
    text-decoration: none;
    margin-right: 30px;
    font-size: 13.5px;
    cursor: pointer;

    &:first-of-type {
        margin-left: 0;
    }

    &:last-of-type {
        margin-right: 0;
    }
`;

export const Group = styled.div`
    display: flex;
    align-items: center;
`;
interface PictureProps {
    src: string;
}

export const Picture = styled.button<PictureProps>`
    background: url(${({src}) => src});
    background-size: contain;
    border: 0;
    width: 32px;
    padding: 8px 15px;
    height: 32px;
    cursor: pointer;
`;

export const Dropdown = styled.div`
    display: none;
    background-color: black;
    position: absolute;
    padding: 10px;
    width: 100px;
    top: 32px;
    right: 10px;
    ${Group}:last-of-type ${Link} {
        cursor: pointer;
    }

    ${Group} {
        margin-bottom: 10px;

        &:last-of-type {
            margin-bottom: 0;
        }

        button {
            margin-right: 10px;
        }

        ${Link}, ${Picture}{
            cursor: default;
        }

        p {
            font-size: 12px;
            margin-bottom: 0;
            margin-top: 0;
        }
    }
`;

export const Profile = styled.div`
    display: flex;
    align-items: center;
    marign-left: 20px;
    position: relative;

    button {
        cursor: pointer;
    }

    &:hover > ${Dropdown}{
        display: flex;
        flex-direction: column;
    }
`;


export const Search = styled.div`
    display: flex;
    align-item: center;

    svg {
        color: white;
        cursosr: pointer;
    }

    @media (max-width: 700px){
        display: none;
    }
`;

export const SearchIcon = styled.button`
    cursor: pointer;
    background-color: transparent;
    border: 0;
    margin-right: 6px;

    img {
        filter: brightness(0) invert(1);
        width: 16px;
    }
`;

interface SearchProps {
    value: any;
    active: any;
}
export const SearchInput = styled.input<SearchProps>`

    background-color: #444444;
    color: white;
    border: 1px solid white;
    transition : width 0.4s;
    height: 30px;
    font-size: 14px;
    margin-left: ${({active}) => active === true ? '10px' : '0'};
    margin-right: ${({active}) => active === true ? '10px' : '0'};
    padding: ${({active}) => active === true ? '10px' : '0'};
    opacity: ${({active}) => active === true ? '1' : '0'};
    width: ${({active}) => active === true ? '200px' : '0'};
`;

export const PlayButton = styled.button`
    box-shadow: 0 0.6vw 1vw -0.4vw rgba(0, 0, 0, .35);
    background-color: white;
    color: black;
    border-width: 0;
    padding: 10px 20px;
    border-radius: 5px;
    display: flex;
    justify-content: space-evenly;
    width: 50%;
    max-width: 180px;
    font-size: 20px;
    font-weight: bold;
    line-height: 30px;
    margin-top: 30px;
    cursor: pointer;
    transition: background-color .25s;
    &: hover {
        background-color: rgba(255,255,255,0.75); 
    }

`;

export const GroupPlayButton = styled(Group)`
    justify-content: space-between;
`;
export const PlayIcon = styled(FaPlay)`
    color: black;
    font-size: 30px;
`;