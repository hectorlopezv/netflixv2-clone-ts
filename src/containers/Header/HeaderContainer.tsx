/* eslint-disable react/jsx-pascal-case */
import React from 'react';
import Header, {Header_Frame, Header_Logo, Header_ButtonLink } from '../../components/Header';

import *  as ROUTES from '../../Routes';
import logo from './logo.svg';
export interface HeaderContainerProps {
    src?: string;
}
 
const HeaderContainer: React.FC<HeaderContainerProps> = ({children, ...props}) => {
    return ( 
        <Header bg = {true} {...props}>
            <Header_Frame>
                <Header_Logo to={ROUTES.HOME} alt="Netflix" src={logo} />
                <Header_ButtonLink to={ROUTES.SIGN_IN}> Sign In</Header_ButtonLink>
            </Header_Frame>
            {children}
        </Header>
     );
}
 
export default HeaderContainer;