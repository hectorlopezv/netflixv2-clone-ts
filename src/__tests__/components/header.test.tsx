/* eslint-disable react/jsx-pascal-case */
import React from 'react';
import { render, fireEvent, getByTestId, getByAltText } from '@testing-library/react';
import Header, {Header_Text, Header_Feature, Header_FeatureCallOut,
    Header_Frame, Header_Logo, Header_TextLink, Header_Group
    ,Header_Profile,Header_Picture, Header_Dropdown, Header_Search,
    Header_PlayButton, Header_ButtonLink
} from '../../components/Header';
import { BrowserRouter as Router } from 'react-router-dom';
import *  as ROUTES from '../../Routes';


//user-event recommended over fireEvent!

jest.doMock('react-router-dom');


describe('Header Components', () => {

    it('Renders Header Component <Header />', () => {
        const {container, getByText, getByTestId} = render(
            <Router>
                <Header bg = {true} >
                    <Header_Frame>
                        <Header_Logo to={ROUTES.HOME} alt="Netflix" src={"./logo.svg"} />
                        
                            <Header_ButtonLink to={ROUTES.SIGN_IN}> Sign In</Header_ButtonLink>
                    </Header_Frame>
                </Header>
            </Router>

        );

        expect(getByText('Sign In')).toBeTruthy();
        expect(getByTestId('header-bg')).toBeTruthy();
        expect(container.firstChild).toMatchSnapshot();
    });

    it('renders full <HEADER /> component', () => {

        const {container, getByText, getByTestId} = render(
            <Router>
            <Header src="joker1" dontShowSmallVp={true}>

            <Header_Frame>
                <Header_Group>
                    <Header_Logo to={ROUTES.HOME} alt="Netflix" src={"./logo.svg"} />
                    <Header_TextLink >Series</Header_TextLink>
                    <Header_TextLink >Films</Header_TextLink> 
                </Header_Group>

                <Header_Group>
                    <Header_Search searchTerm={"Joker"} setsearchTerm={() => {}} />
                    <Header_Profile>
                        <Header_Picture src={"/images/karl.png"} />
                        <Header_Dropdown>

                            <Header_Group>
                                <Header_Picture  src={"/images/karl.png"} />
                                <Header_TextLink>karln dawn</Header_TextLink>
                            </Header_Group>

                            <Header_Group>
                                <Header_TextLink onClick={() => {}}>Sign Out</Header_TextLink>
                            </Header_Group>

                        </Header_Dropdown>


                    </Header_Profile>
                </Header_Group>


            </Header_Frame>

                <Header_Feature>
                    <Header_FeatureCallOut>
                        Watch Joker Now
                    </Header_FeatureCallOut>
                    <Header_Text>
                        Forever alone in a crowd.
                    </Header_Text>
                
                    <Header_Group>
                        <Header_PlayButton>Play</Header_PlayButton>
                    </Header_Group>

                </Header_Feature>
            </Header>

            </Router>

        );

        
        expect(getByTestId('seach-input')).toBeTruthy();
        expect(getByTestId('seach-input').value).toBe('Joker');

        fireEvent.change(getByTestId('seach-input'), {target: {value: 'Simpsons'},
        });
        fireEvent.click(getByTestId('seach-click'));

        expect(getByText('Series')).toBeTruthy();
        expect(getByText('Films')).toBeTruthy();
        expect(getByText('karln dawn')).toBeTruthy();
        expect(getByText('Watch Joker Now')).toBeTruthy();
        expect(getByText('Forever alone in a crowd.')).toBeTruthy();
        expect(getByText('Play')).toBeTruthy();

        expect(container.firstChild).toMatchSnapshot();
    });

});