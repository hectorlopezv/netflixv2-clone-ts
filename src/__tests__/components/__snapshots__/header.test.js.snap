// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Components Renders Header Component <Header /> 1`] = `
<div
  class="Header__Background-sc-3orlxz-3 gELmNw"
  data-testid="header-bg"
>
  <div
    class="Header__Container-sc-3orlxz-0 eJgkia"
  >
    <a
      href="/"
    >
      <img
        alt="Netflix"
        class="Header__Logo-sc-3orlxz-1 hLarDa"
        src="./logo.svg"
      />
    </a>
    <a
      class="Header__Header_ButtonLink-sc-3orlxz-4 hpBLda"
      href="/signin"
      style="text-decoration: none;"
    >
       Sign In
    </a>
  </div>
</div>
`;

exports[`Header Components Renders Header Component <Header /> 2`] = `
<div
  class="Header__Background-sc-3orlxz-3 gELmNw"
  data-testid="header-bg"
>
  <div
    class="Header__Container-sc-3orlxz-0 eJgkia"
  >
    <a
      href="/"
    >
      <img
        alt="Netflix"
        class="Header__Logo-sc-3orlxz-1 hLarDa"
        src="./logo.svg"
      />
    </a>
    <a
      class="Header__Header_ButtonLink-sc-3orlxz-4 hpBLda"
      href="/signin"
      style="text-decoration: none;"
    >
       Sign In
    </a>
  </div>
</div>
`;

exports[`Header Components renders full <HEADER /> component 1`] = `
<div
  class="Header__Background-sc-3orlxz-3 fzVgVC"
  data-testid="header-bg"
  src="joker1"
>
  <div
    class="Header__Container-sc-3orlxz-0 eJgkia"
  >
    <div
      class="Header__Group-sc-3orlxz-9 kcdRoF"
    >
      <a
        href="/"
      >
        <img
          alt="Netflix"
          class="Header__Logo-sc-3orlxz-1 hLarDa"
          src="./logo.svg"
        />
      </a>
      <p
        class="Header__Link-sc-3orlxz-8 dkZWnF"
      >
        Series
      </p>
      <p
        class="Header__Link-sc-3orlxz-8 dkZWnF"
      >
        Films
      </p>
    </div>
    <div
      class="Header__Group-sc-3orlxz-9 kcdRoF"
    >
      <div
        class="Header__Search-sc-3orlxz-13 iFrXpU"
      >
        <button
          class="Header__SearchIcon-sc-3orlxz-14 izWlhZ"
        >
          <img
            alt="Search"
            src="/images/icons/search.png"
          />
        </button>
        <input
          class="Header__SearchInput-sc-3orlxz-15 eLamtx"
          data-testid="seach-input"
          placeholder="Search films and series"
          value="Joker"
        />
      </div>
      <div
        class="Header__Profile-sc-3orlxz-12 gyLCcS"
      >
        <button
          class="Header__Picture-sc-3orlxz-10 ckmpfO"
          src="/images/users//images/karl.png.png"
        />
        <div
          class="Header__Dropdown-sc-3orlxz-11 fbEVHg"
        >
          <div
            class="Header__Group-sc-3orlxz-9 kcdRoF"
          >
            <button
              class="Header__Picture-sc-3orlxz-10 ckmpfO"
              src="/images/users//images/karl.png.png"
            />
            <p
              class="Header__Link-sc-3orlxz-8 dkZWnF"
            >
              karln dawn
            </p>
          </div>
          <div
            class="Header__Group-sc-3orlxz-9 kcdRoF"
          >
            <p
              class="Header__Link-sc-3orlxz-8 dkZWnF"
            >
              Sign Out
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="Header__Container-sc-3orlxz-0 Header__Feature-sc-3orlxz-5 eJgkia lhezVz"
  >
    <h2
      class="Header__FeatureCallOut-sc-3orlxz-7 byYlgc"
    >
      Watch Joker Now
    </h2>
    <p
      class="Header__Text-sc-3orlxz-6 jUjAin"
    >
      Forever alone in a crowd.
    </p>
    <div
      class="Header__Group-sc-3orlxz-9 kcdRoF"
    >
      <button
        class="Header__PlayButton-sc-3orlxz-16 doDfKT"
      >
        Play
      </button>
    </div>
  </div>
</div>
`;
